mlflow/pyfunc/__init__.py

```diff
@@ -206,6 +206,100 @@
+
+- **Environment Dependency**: cloudpickle does not capture the full execution environment, so in
+    MLflow we must pass ``pip_requirements``, ``extra_pip_requirements``, or an ``input_example``,
+    the latter of which is used to infer environment dependencies.
+
+- **Object Support**: cloudpickle does not serialize objects outside of the Python data model.
+    Some relevant examples include raw files and database connections. If your program depends on
+    these, be sure to log ways to reference these objects along with your model.
+
+python_function
+###############
+If you're looking to serialize a simple python function without additional dependent methods, you
+can simply log a predict method via the keyword argument ``python_model``.
+
+.. code-block:: python
+    :caption: python_function Example
+
+    import mlflow
+    import pandas as pd
+
+    # Define a simple function to log
+    def predict(model_input):
+        return model_input.apply(lambda x: x * 2)
+
+    # Save the function as a model
+    with mlflow.start_run():
+        mlflow.pyfunc.log_model("model", python_model=predict, pip_requirements=["pandas"])
+        run_id = mlflow.active_run().info.run_id
+
+    # Load the model from the tracking server and perform inference
+    model = mlflow.pyfunc.load_model(f"runs:/{run_id}/model")
+    x_new = pd.Series([1,2,3])
+
+    print(f"Prediction:\n\t{model.predict(x_new)}")
+
+
+PythonModel
+###############
+If you're looking to serialize a more complex object, for instance a class that handles
+preprocessing, complex prediction logic, or custom serialization, you should use the
+``PythonModel`` class. MLflow has tutorials on building custom PyFunc models, as shown
+`here <https://mlflow.org/docs/latest/traditional-ml/creating-custom-pyfunc/index.html>`_,
+so instead of duplicating that information, in this example we'll recreate the above functionality
+to highlight the differences. Note that this PythonModel implementation is overly complex and
+should be a python_function instead.
+
+.. code-block:: python
+    :caption: PythonModel Example
+
+    import mlflow
+    import pandas as pd
+
+    class MyModel(mlflow.pyfunc.PythonModel):
+        def predict(self, context, model_input, params=None):
+            return model_input.apply(lambda x: x * 2)
+
+    # Save the function as a model
+    with mlflow.start_run():
+        mlflow.pyfunc.log_model("model", python_model=MyModel(), pip_requirements=["pandas"])
+        run_id = mlflow.active_run().info.run_id
+
+    # Load the model from the tracking server and perform inference
+    model = mlflow.pyfunc.load_model(f"runs:/{run_id}/model")
+    x_new = pd.Series([1, 2, 3])
+
+    print(f"Prediction:\n\t{model.predict(x_new)}")
+
+The primary difference between the this implementation and the python_function implementation above
+is that the predict method is wrapped with a class, has the ``self`` parameter,
+and has the ``params`` parameter that defaults to None.
```

Comment:
nit: Can we add `params` to the function-based model above? It might be misleading that the acceptance of `params` is difference between those two interface.

Reply:
When adding logging a function-based model with params=None, we get the following error. [Source code](https://github.com/mlflow/mlflow/blob/eec816f06019fa6d568dea434a56cad996abe6ee/mlflow/pyfunc/__init__.py#L1822)
```
mlflow.exceptions.MlflowException: When `python_model` is a callable object, it must accept exactly one argument. Found 2 arguments.
```
I'll add a note that function-based models don't support params.

Reply:
Oooh I didn't know that! Yeah let's add that note explicitly.